packages/aws-cdk/test/integ/cli/bootstrapping.integtest.ts

```diff
@@ -1,52 +1,175 @@
-import { cloudFormation, deleteStacks } from './aws-helpers';
-import { cdk, cleanupOldStacks, fullStackName, prepareAppFixture } from './cdk-helpers';
+import { cloudFormation, deleteStacks, s3 } from './aws-helpers';
+import { cdk, cdkDeploy, cleanupOldStacks, deleteableStacks, fullStackName, prepareAppFixture } from './cdk-helpers';
 
-jest.setTimeout(600 * 1000);
+jest.setTimeout(600_000);
+
+// Needs to not start with the ACTUAL fullstack name otherwise cleanupOldStacks will pick it up.
+const BOOTSTRAP_STACK_PREFIX = `toolkit-${fullStackName('stack')}`;
+const BUCKETS_TO_DELETE = new Array<string>();
+const QUALIFIER = randomString();
 
 beforeAll(async () => {
   await prepareAppFixture();
 });
 
 beforeEach(async () => {
   await cleanupOldStacks();
+  BUCKETS_TO_DELETE.splice(0, BUCKETS_TO_DELETE.length);
 });
 
 afterEach(async () => {
   await cleanupOldStacks();
+  await cleanupBootstrapStacks();
+
+  // We might have lost some buckets by upgrading the bootstrap stack. Be
+  // sure to clean everything.
+  for (const bucket of BUCKETS_TO_DELETE) {
+    await deleteBucket(bucket);
+  }
 });
 
 test('can bootstrap without execution', async () => {
-  const bootstrapStackName = fullStackName('toolkit-stack-1');
+  const bootstrapStackName = `${BOOTSTRAP_STACK_PREFIX}-1`;
 
-  await cdk(['bootstrap', '-v',
+  await cdk(['bootstrap',
     '--toolkit-stack-name', bootstrapStackName,
     '--no-execute']);
+
+  const resp = await cloudFormation('describeStacks', {
+    StackName: bootstrapStackName,
+  });
+
+  expect(resp.Stacks?.[0].StackStatus).toEqual('REVIEW_IN_PROGRESS');
+});
+
+test('upgrade legacy bootstrap stack to new bootstrap stack while in use', async () => {
+  const bootstrapStackName = `${BOOTSTRAP_STACK_PREFIX}-update-in-place`;
+
+  const legacyBootstrapBucketName = `aws-cdk-bootstrap-integ-test-legacy-bckt-${randomString()}`;
+  const newBootstrapBucketName = `aws-cdk-bootstrap-integ-test-v2-bckt-${randomString()}`;
+  BUCKETS_TO_DELETE.push(legacyBootstrapBucketName, newBootstrapBucketName);
+
+  // Legacy bootstrap
+  await cdk(['bootstrap',
+    '--toolkit-stack-name', bootstrapStackName,
+    '--bootstrap-bucket-name', legacyBootstrapBucketName]);
+
+  // Deploy stack that uses file assets
+  await cdkDeploy('lambda', {
+    options: ['--toolkit-stack-name', bootstrapStackName],
+  });
+
+  // Upgrade bootstrap stack to "new" style
+  await cdk(['bootstrap',
+    '--toolkit-stack-name', bootstrapStackName,
+    '--bootstrap-bucket-name', newBootstrapBucketName,
+    '--qualifier', QUALIFIER], {
+    modEnv: {
+      CDK_NEW_BOOTSTRAP: '1',
+    },
+  });
+
+  // (Force) deploy stack again
+  // --force to bypass the check which says that the template hasn't changed.
+  await cdkDeploy('lambda', {
```

Comment:
I think this isn't testing the same thing the old test did.

We should deploy a new version of the app, with an asset that's changed, _but_ still using the old asset resolution (the one that used parameters instead of hardcoding the paths).

I think this will not update the asset to the (new) bucket, because nothing in the template has changed (the CFN deployment will be a no-op).

Reply:
That's why I'm passing `--force`.

Reply:
I don't think it will be the same...? The CloudFormation deployment will not happen, so it's not actually testing whether the assets are put in the new bucket correctly?

I guess you're saying that if the asset upload failed, the `cdkDeploy` command would fail anyway, so there's no point in doing a CloudFormation deployment?

If so, OK, I can buy that, although personally, I would still do a CloudFormation deployment. I'm not blocking the PR on it though.